ui/src/main.ts

```diff
@@ -3,10 +3,20 @@ import {platformBrowserDynamic} from '@angular/platform-browser-dynamic';
 
 import {AppModule} from './app/app.module';
 import {environment} from './environments/environment';
+import {ConfigLoaderService} from "./environments/config-loader.service";
 
 if (environment.production) {
   enableProdMode();
 }
 
-platformBrowserDynamic().bootstrapModule(AppModule)
-  .catch(err => console.log(err));
+const configLoader = new ConfigLoaderService();
+
+configLoader.getEnvironmentConfig()
+  //TODO(Rex): Show err with snackBar(optional) and stop app peacefully
+  .catch(err => console.log(err))
+  .then(() => {
+    platformBrowserDynamic([{
+      provide: ConfigLoaderService, useValue: configLoader
```

Comment:
What is the difference between including `ConfigLoaderService` as a provider here instead of in app.module.ts?

Reply:
Good question! Normally we put the service class name in the `app.module.ts` so that the provider will be responsible for creating singleton/instantiating the service. But in this case, we want the LoaderService's promise to be executed before all other services' initialization, which means we have to instantiate the service singleton by ourselves, make it store the environment variables which are previously stored in envrionments.ts files, and pass it to the module providers. So providers will try to use this specific singleton (by `useValue`) instead of creating an empty new one. I tried also include the class in `app.module.ts` at the same time, but it turned out that will shadow this one. What do you think?

Reply:
I think that makes sense and this is the right place to put it 👍

Reply:
Thank you for the explanation! I think the way you have it now makes sense.